Adds test and deploy workflows

p-galligan opened the pull request:
Adds workflows for testing and deployment

p-galligan committed:
Add preliminary cicd

Adds preliminary ci/cd workflows

p-galligan committed:
update cicd

Updates cicd workflow to bypass TTY

p-galligan committed:
update requirements.txt

Re-runs `pip-compile` to update requirements.txt

p-galligan committed:
Update ci workflow

Updates CI workflow to remove TTY

p-galligan committed:
update cicd

Updates checkout version

p-galligan committed:
cicd workflow updates

Testing deploy job

p-galligan committed:
force deploy

comments out if to force deploy

p-galligan committed:
testing

Tests version of s3 deploy

p-galligan committed:
remove codedeploy

Commenting out codedeploy for testing

p-galligan committed:
fix variable

fix variable

p-galligan committed:
fix variable name

Fixes variables

p-galligan committed:
try different approach

try a different approach to push to s3

p-galligan committed:
remove outdated code

Removes outdated code sections

p-galligan committed:
update branch target

Targets different branch for deploy

p-galligan committed:
Revert test

testing environment variable

p-galligan committed:
Fix typo

Fixes a typo in workflow

p-galligan committed:
secrets variable test

Testing whether secrets are encoded

p-galligan committed:
fix typo

fixes a typo in secret test

p-galligan committed:
Remove test lines

Removes testing lines

p-galligan committed:
update deploy workflow

Updates the deploy workflow to fix CodeDeploy deployment.

p-galligan committed:
environment test

updates workflow to test environment substitution

p-galligan committed:
testing

test variables

p-galligan committed:
test envsubst

Testing environment subsitution

p-galligan committed:
Deploy to s3

Uncomment out the S3 deploy.

p-galligan committed:
add files for envsubst

Add files for envsubst action

p-galligan committed:
Add files for envsubst

Lists files for envsubst.

p-galligan committed:
remove unnecessary code

Removes duplicate code

p-galligan committed:
rename files

Add step to rename files

p-galligan committed:
rename file test

test to rename files

p-galligan committed:
Allow codedeploy

Uncomments codedeploy deployment creation

p-galligan committed:
Change variable calls

Changes workflows to use environment variables for codedeploy

p-galligan committed:
Change environment choice

Change workflow action to use the environment based on the github ref name

p-galligan committed:
Update setup-python version

Updates the setup-python version in CI/CD

p-galligan committed:
test role assumption

Adds assume role to deploy

p-galligan committed:
fixe environmet

Forgot to set environment for testing

p-galligan committed:
remove session

Tests by removing session name

p-galligan committed:
add session back

Adds session name back

p-galligan committed:
Check aws user

add step to check sts

p-galligan committed:
use cp instead of sync

Uses cp instead of sync for s3 upload

p-galligan committed:
Add recursive to workflow

Adds recursive flag to s3 upload

p-galligan committed:
remove session name

Testing without session name

p-galligan committed:
skip session tagging

Skips session tagging in configuring aws

p-galligan committed:
switch roles for codedeploy

Adds a role switch for CodeDeploy creation

p-galligan committed:
Updates workflow environment

Updates workflow environment after testing has finished.

p-galligan committed:
test new roles

Testing new roles

p-galligan committed:
testing with different user

Testing cd workflow with different user

p-galligan committed:
testing permissions

Testing S3 permissions

p-galligan committed:
testing ACL

Testing s3 with ACL

p-galligan committed:
confirm caller identity

confirm identity of codedeploy role

p-galligan committed:
remove acl

Removes push with acl to s3

p-galligan committed:
remove caller identity check

Removes check for caller identity from GHA workflow

p-galligan committed:
tweak workflow

Tweaks the workflow to run only when wanted.

p-galligan committed:
update workflows

Breaks test and deploy into two different workflows. Removes Travis file.

p-galligan committed:
fix typo

fixes indentation typo

p-galligan committed:
add on PR for testing

adds run on PR for testing

p-galligan committed:
testing

testing environments

p-galligan committed:
testing old workflow

tests old workflow by adding back in

p-galligan committed:
fixing that

fixes test

p-galligan committed:
Merge branch 'development' into workflows

p-galligan committed:
update workflows

splits workflows

p-galligan committed:
Revert "update workflows"

This reverts commit 79ac5246b394694fc93be7d54a3688848a5210c6.

p-galligan committed:
Merge branch 'workflows' of https://github.com/RockefellerArchiveCenter/gemini into workflows

p-galligan committed:
testing workflows

updates workflows

p-galligan committed:
testing deploy

Testing separate deploy actions

p-galligan committed:
comment out conditional

comments out a conditional for testing deploy

p-galligan committed:
Fix environment setting

Fixes environment in deploy, removes on PR run, and adds conditional back in.

helrond commented on .github/workflows/deploy.yml:
```diff
@@ -0,0 +1,102 @@
+jobs:
+  deploy:
+    runs-on: ubuntu-latest
+    if: github.ref == 'refs/heads/main' || github.ref == 'refs/heads/development'
```
Do we need this still?

helrond replied:
Also, if we do, it should be `refs/heads/base` not `refs/heads/main`

helrond commented on .github/workflows/deploy.yml:
```diff
@@ -0,0 +1,102 @@
+      - name: Create deployment zip
+        run: sudo deploy_scripts/make_zip_django.sh $DEPLOY_ZIP_DIR $DEPLOY_ZIP_NAME
+
+      - name: Configure AWS Credentials
```
I know there's a reason we're assuming two different roles but I'm blanking on what it is right now and if this is a temporary workaround or how we expect things to be long-term.

p-galligan replied:
Yeah, so this is a temporary workaround. The dev CodeDeploy role doesn't have permissions for prod CodeDeploy. 

As discussed, the ideal would be to have all dev resources (CodeDeploy) on the same account. I just went for the route of adding permissions to assume the prod role for CodeDeploy. I opted for the dev deployment role that already exists for the S3 push. 

Ideally, we'd just need a single role. I _could_ play with adding prod permissions to the dev CodeDeploy role if that's preferable.

helrond replied:
I'm just trying to think about making our lives easier in the future. We're going to have to update the role ARN anyway, but is there a way we can avoid having to update the workflow YAML as well?

p-galligan replied:
Ok, I just updated the Role ARN for dev to match permissions for the prod one.

Unfortunately, I had forgotten what the original issue was. As far as I can tell, there's no way to create a code deploy across different accounts because it won't find the application you need.

I don't see any way to specify which account to create a CodeDeploy deployment in the [documentation](https://github.com/RockefellerArchiveCenter/gemini/actions/runs/8993118588/job/24704293632).

You can see the error message [here](https://github.com/RockefellerArchiveCenter/gemini/actions/runs/8993118588/job/24704293632).

helrond replied:
OK great, as discussed we can continue to use the prod role, which has all the necessary permissions.

helrond commented on .github/workflows/tests.yml:
```diff
@@ -0,0 +1,67 @@
+
+jobs:
+  build:
+    if: github.ref != 'refs/heads/base' || github.ref != 'refs/heads/development'
```
I think this can be removed.

helrond commented on .github/workflows/tests.yml:
```diff
@@ -0,0 +1,67 @@
+    if: github.ref != 'refs/heads/base' || github.ref != 'refs/heads/development'
+    runs-on: ubuntu-latest
+    environment:
+      name: development
```
I know we're only running tests on the development branch but does it make sense to use a variable here in case we ever change that pattern?

helrond reviewed (changes requested):
A few small things.

p-galligan committed:
Remove job conditionals

Removes job conditionals from tests and deploys.

Adds a repository variable for targeting environments on test.

p-galligan committed:
Update deploy workflow

tests new role permissions for deploy

p-galligan committed:
fix environment

fixes environment variable

p-galligan committed:
revert deploy workflow changes

reverts changes to the deploy workflow

p-galligan committed:
test deploy workflow

tests deploy workflow to use a single role in production account

p-galligan committed:
update deploy

Restores environment and run conditionals for deploys after testing.

p-galligan committed:
testing

testing deploy again

p-galligan committed:
update workflows

remove unnecessary lines from deploy
